dowhy/causal_refuters/bootstrap_refuter.py

```diff
@@ -1,184 +1,208 @@
+        else:    
+            self.logger.info("INFO: The chosen variables are: " +
+                            ",".join(self._chosen_variables))
+
+        if self._probability_of_change is None:
+            if self._noise > 1:
+                self.logger.error("Error in using noise:{} for Binary Flip. The value is greater than 1".format(self._noise))
+                raise ValueError("The value for Binary Flip cannot be greater than 1")
+            else:
+                self._probability_of_change = self._noise
+        elif self._probability_of_change > 1:
+            self.logger.error("The probability of flip is: {}, However, this value cannot be greater than 1".format(self._probability_of_change))
+            raise ValueError("Probability of Flip cannot be greater than 1")
+
+    
+    def refute_estimate(self, *args, **kwargs):
+        if self._sample_size > len(self._data):
+                self.logger.warning("The sample size is larger than the population size")
+
+        sample_estimates = np.zeros(self._num_simulations)
+        self.logger.info("Refutation over {} simulated datasets of size {} each"
+                         .format(self._num_simulations
+                         ,self._sample_size )
+                        ) 
+        
+        for index in range(self._num_simulations):
+            if self._random_state is None:
+                new_data = resample(self._data, 
+                                n_samples=self._sample_size )
+            else:
+                new_data = resample(self._data,
+                                    n_samples=self._sample_size,
+                                    random_state=self._random_state )
+
+
+            if self._chosen_variables is not None:
+                for variable in self._chosen_variables:
+                    
+                    if ('float' or 'int') in new_data[variable].dtype.name:
+                        scaling_factor = new_data[variable].std() 
+                        new_data[variable] += np.random.randn(self._sample_size) * self._noise * scaling_factor
+                    
+                    elif 'bool' in new_data[variable].dtype.name:
+                        probs = np.random.uniform(0, 1, self._sample_size )
+                        new_data[variable] = np.where(probs < self._probability_of_change, 
+                                                        np.logical_not(new_data[variable]), 
+                                                        new_data[variable]) 
+                    
+                    elif 'category' in new_data[variable].dtype.name:
+                        self.categories = new_data[variable].unique()
+                        # Find the set difference for each row
+                        changed_data = new_data[variable].apply(self.set_diff)
+                        # Choose one out of the remaining
+                        changed_data = changed_data.apply(self.choose_random)
```

Comment:
rather than choose_random which is single line function, a lambda function is better here. Also more readable. 
lambda x: random.choice(x)